### PR TITLE
Reject package with embeddedReadme file with featureFlag

### PR DESCRIPTION
**Pull request by lyndaidaii**
Summary of the changes:

* add feature flag to enable embedded readme within package 
* separate validation before uploading packages

additional work need to be done in terms of separating validation in validation service

Addresses https://github.com/nuget/engineering/issues/2342

### CODE REVIEW

File: src/NuGetGallery/Services/IPackageMetadataValidationService.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+namespace NuGetGallery
+{
+    public interface IPackageMetadataValidationService
+    {
+        Task<PackageValidationResult> ValidateMetadatabeforeUploadAsyn(
+            PackageArchiveReader nuGetPackage,
+            PackageMetadata packageMetadata,
+            User currentUser);
+
```
Comment (lyndaidaii) [inline src/NuGetGallery/Services/IPackageMetadataValidationService.cs:14]: delete line here

File: src/NuGetGallery/Services/PackageMetadataValidationService.cs [hunk 2]
```diff
@@ -0,0 +1,754 @@
...
+using NuGet.Versioning;
+using NuGetGallery.Diagnostics;
+using System.IO;
+using NuGet.Packaging.Licenses;
+using System.Xml.Linq;
+
+namespace NuGetGallery
+{
+    public class PackageMetadataValidationService : IPackageMetadataValidationService
+
```
Comment (lyndaidaii) [inline src/NuGetGallery/Services/PackageMetadataValidationService.cs:23]: delete line here

File: src/NuGetGallery/Services/PackageMetadataValidationService.cs [hunk 3]
```diff
@@ -0,0 +1,754 @@
...
+        private const string AllowedLicenseVersion = "1.0.0";
+        private const string Unlicensed = "UNLICENSED";
+
+        private readonly IPackageService _packageService;
+        private readonly IAppConfiguration _config;
+        private readonly ITelemetryService _telemetryService;
+        private readonly IDiagnosticsSource _trace;
+        private readonly IFeatureFlagService _featureFlagService;
+
+
```
Comment (lyndaidaii) [inline src/NuGetGallery/Services/PackageMetadataValidationService.cs:75]: delete line here

File: src/NuGetGallery/Services/PackageMetadataValidationService.cs [hunk 4]
```diff
@@ -0,0 +1,754 @@
...
+            result = await CheckReadmeMetadataAsync(nuGetPackage, warnings, currentUser);
+            if (result != null)
+            {
+                return result;
+            }
+
+            return PackageValidationResult.AcceptedWithWarnings(warnings);
+        }
+        private async Task<PackageValidationResult> CheckLicenseMetadataAsync(PackageArchiveReader nuGetPackage, List<IValidationMessage> warnings, User user)
```
Comment (lyndaidaii) [inline src/NuGetGallery/Services/PackageMetadataValidationService.cs:162]: add a newline here

File: src/NuGetGallery/Services/PackageMetadataValidationService.cs [hunk 5]
```diff
@@ -0,0 +1,754 @@
...
+                        Strings.UploadPackage_FileTooLong,
+                        Strings.UploadPackage_ReadmeFileType,
+                        MaxAllowedReadmeLengthForUploading.ToUserFriendlyBytesLabel()));
+            }
+
+            if (!await IsStreamLengthMatchesReportedAsync(nuGetPackage, readmeFilePath, readmeFileEntry.Length))
+            {
+                return PackageValidationResult.Invalid(Strings.UploadPackage_CorruptNupkg);
+            }
+            //check the supported the md format
```
Comment (lyndaidaii) [inline src/NuGetGallery/Services/PackageMetadataValidationService.cs:472]: more validation about supported md format need to be added later. I will create new ticket to track this issue to address later once we define the details about supported md

File: src/NuGetGallery/Services/PackageMetadataValidationService.cs [hunk 6]
```diff
@@ -0,0 +1,754 @@
...
+
+            if (!isPng && !isJpg)
+            {
+                return PackageValidationResult.Invalid(Strings.UploadPackage_UnsupportedIconImageFormat);
+            }
+
+            return null;
+        }
+
+        private async Task<PackageValidationResult> CheckReadmeMetadataAsync(PackageArchiveReader nuGetPackage, List<IValidationMessage> warnings, User user)
```
Comment (lyndaidaii) [inline src/NuGetGallery/Services/PackageMetadataValidationService.cs:416]: new function added for readme, need add more validation later to check if we are supporting format for markdown

File: src/NuGetGallery.Services/Configuration/FeatureFlagService.cs [hunk 7]
```diff
@@ -37,6 +37,7 @@ public class FeatureFlagService : IFeatureFlagService
         private const string Get2FADismissFeedback = GalleryPrefix + "Get2FADismissFeedback";
         private const string UsabillaOnEveryPageFeatureName = GalleryPrefix + "UsabillaEveryPage";
         private const string PackageRenamesFeatureName = GalleryPrefix + "PackageRenames";
+        private const string UploadEmbeddedReadMeFeatureName = GalleryPrefix + "UploadEmbeddedReadMeFeatureName";
```
Comment (loic-sharma) [inline src/NuGetGallery.Services/Configuration/FeatureFlagService.cs:40]: Can we change the styling of readme everywhere from `ReadMe` to `Readme`? This is a single word, our previous usage was incorrect.

```suggestion
        private const string UploadEmbeddedReadMeFeatureName = GalleryPrefix + "UploadEmbeddedReadmeFeatureName";
```

File: src/NuGetGallery.Services/Configuration/FeatureFlagService.cs [hunk 8]
```diff
@@ -185,5 +186,10 @@ public bool IsPackageRenamesEnabled(User user)
         {
             return _client.IsEnabled(PackageRenamesFeatureName, user, defaultValue: false);
         }
+
+        public bool IsUploadEmbeddedReadMeEnabled(User user)
```
Comment (loic-sharma) [inline src/NuGetGallery.Services/Configuration/FeatureFlagService.cs:190]:
```suggestion
        public bool IsUploadEmbeddedReadmeEnabled(User user)
```

File: src/NuGetGallery.Services/PackageManagement/PackageService.cs [hunk 9]
```diff
@@ -711,6 +712,23 @@ private static EmbeddedLicenseFileType GetEmbeddedLicenseType(string licenseFile
             throw new ArgumentException($"Invalid file name: {licenseFileName}");
         }
 
+        private static EmbeddedReadmeFileType GetEmbeddedReadmeType(PackageMetadata packageMetadata)
+        {
+            const string MarkdownFileExtension = ".md";
+            if (packageMetadata.ReadmeFile == null) {
```
Comment (loic-sharma) [inline src/NuGetGallery.Services/PackageManagement/PackageService.cs:718]: Nit:

```suggestion
            if (packageMetadata.ReadmeFile == null)
            {
```

File: src/NuGetGallery/Services/IPackageMetadataValidationService.cs [hunk 10]
```diff
@@ -0,0 +1,16 @@
+﻿using NuGet.Packaging;
+using NuGet.Services.Entities;
+using NuGetGallery.Packaging;
+using System.Threading.Tasks;
+
+namespace NuGetGallery
+{
+    public interface IPackageMetadataValidationService
+    {
+        Task<PackageValidationResult> ValidateMetadatabeforeUploadAsyn(
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/IPackageMetadataValidationService.cs:10]:
```suggestion
        Task<PackageValidationResult> ValidateMetadataBeforeUploadAsync(
```

Comment (loic-sharma) [inline src/NuGetGallery/Services/IPackageMetadataValidationService.cs:10]: Please add a quick comment explaining this method

File: tests/NuGetGallery.Facts/NuGetGallery.Facts.csproj [hunk 11]
```diff
@@ -1,5 +1,14 @@
...
 <Project ToolsVersion="15.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <SSDTUnitTestPath Condition="'$(SSDTUnitTestPath)' == ''">$(VsInstallRoot)\Common7\IDE\Extensions\Microsoft\SQLDB</SSDTUnitTestPath>
+  </PropertyGroup>
+  <PropertyGroup>
+    <SSDTPath Condition="'$(SSDTPath)' == ''">$(VsInstallRoot)\Common7\IDE\Extensions\Microsoft\SQLDB\DAC\130</SSDTPath>
+  </PropertyGroup>
+  <PropertyGroup>
+    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">10.0</VisualStudioVersion>
+  </PropertyGroup>
```
Comment (loic-sharma) [inline tests/NuGetGallery.Facts/NuGetGallery.Facts.csproj:11]: What's this for? Can we undo these changes?

Comment (loic-sharma) [reply tests/NuGetGallery.Facts/NuGetGallery.Facts.csproj:11]: Same comment for lines 388-402 and 410-414

File: src/NuGetGallery/Services/PackageMetadataValidationService.cs [hunk 12]
```diff
@@ -0,0 +1,754 @@
...
+using System.IO;
+using NuGet.Packaging.Licenses;
+using System.Xml.Linq;
+
+namespace NuGetGallery
+{
+    public class PackageMetadataValidationService : IPackageMetadataValidationService
+
+    {
+        private static readonly IReadOnlyCollection<string> AllowedLicenseFileExtensions = new HashSet<string>
```
Comment (cristinamanum) [inline src/NuGetGallery/Services/PackageMetadataValidationService.cs:25]: Is all this code moved from  [here](https://github.com/NuGet/NuGetGallery/pull/8118/files#diff-b426147e28e1af03bb76fb773d48199e) ?

File: src/NuGetGallery/Services/PackageMetadataValidationService.cs [hunk 13]
```diff
@@ -0,0 +1,754 @@
...
+            }
+
+            var readmeFileExtension = Path.GetExtension(readmeFilePath);
+            if (!AllowedReadmeFileExtensions.Contains(readmeFileExtension, StringComparer.OrdinalIgnoreCase))
+            {
+                var result = PackageValidationResult.Invalid(
+                    string.Format(
+                        Strings.UploadPackage_InvalidReadmeFileExtension,
+                        readmeFileExtension,
+                        string.Join(", ", AllowedReadmeFileExtensions.Where(x => x != string.Empty).Select(extension => $"'{extension}'"))));
```
Comment (cristinamanum) [inline src/NuGetGallery/Services/PackageMetadataValidationService.cs:454]: Why is the Where needed?

File: tests/NuGetGallery.Facts/Services/PackageMetadataValidationServiceFacts.cs [hunk 14]
```diff
@@ -0,0 +1,1771 @@
...
+                    _nuGetPackage.Object,
+                    GetPackageMetadata(_nuGetPackage),
+                    _currentUser);
+            }
+
+            public async Task RejectPackagesWithEmbeddedIcon() 
+            {
+                _nuGetPackage = GeneratePackageWithUserContent(
+                    readmeFilename: "readme.md",
+                    readmeFileContents: "some readme md");
```
Comment (cristinamanum) [inline tests/NuGetGallery.Facts/Services/PackageMetadataValidationServiceFacts.cs:1362]: What is the connection between the readme and icon?

File: src/NuGetGallery/Services/PackageMetadataValidationService.cs [hunk 15]
```diff
@@ -0,0 +1,754 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using NuGet.Packaging;
```
Comment (cristinamanum) [inline src/NuGetGallery/Services/PackageMetadataValidationService.cs:4]: Out of order.

File: src/NuGetGallery/Services/PackageMetadataValidationService.cs (hunk 12 above)
Comment (lyndaidaii) [reply src/NuGetGallery/Services/PackageMetadataValidationService.cs:25]: basically yes, I added   ` private async Task<PackageValidationResult> CheckReadmeMetadataAsync(PackageArchiveReader nuGetPackage, List<IValidationMessage> warnings, User user)`

and unit test for this function

**Comment by lyndaidaii** [comment]
Make two PR for easy review: 
add feature flag:  https://github.com/NuGet/NuGetGallery/pull/8119
